### PR TITLE
feat: add export-terraform function for exporting KCC resources as Terraform

### PR DESCRIPTION
**Pull request by morgante**
This adds a Terraform export function.

See go/terraform-krm-export for more context.

### CODE REVIEW

**Review by droot — APPROVED** [review]

**Comment by droot** [comment]
@morgante CI is reporting one of the tests to be failing

```
 --- FAIL: TestE2E/../../examples/export-terraform-imperative
```

File: functions/go/export-terraform/go.mod [hunk 1]
```diff
@@ -2,11 +2,11 @@ module github.com/GoogleContainerTools/kpt-functions-catalog/functions/go/export
 
 go 1.16
 
-replace sigs.k8s.io/kustomize/kyaml v0.11.1 => github.com/mengqiy/kustomize/kyaml v0.15.0
+replace github.com/GoogleContainerTools/kpt-functions-catalog/thirdparty/kyaml/fnsdk v0.0.0-20220106201853-53d6f3d583fe => ../../../thirdparty/kyaml/fnsdk
```
Comment (mengqiy) [inline functions/go/export-terraform/go.mod:5]: We got some push back when trying to merge the new SDK in upstream. We are actively working on it to address the concerns.
Meanwhile we want to unblock the OC project. So I have moved the code (github.com/mengqiy/kustomize/kyaml v0.15.0) from my fork to the catalog repo in https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/724.

File: examples/export-terraform-imperative/Kptfile [hunk 2]
```diff
@@ -0,0 +1,5 @@
+apiVersion: kpt.dev/v1
+kind: Kptfile
+metadata:
+  name: export-terraform-imperative
+pipeline: {}
```
Comment (mengqiy) [inline examples/export-terraform-imperative/Kptfile:5]: Is this file necessary?

File: functions/go/export-terraform/main.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	if err := sdk.AsMain(sdk.ResourceListProcessorFunc(Process)); err != nil {
+		os.Exit(1)
+	}
+}
+
+func Process(resourceList *sdk.ResourceList) error {
+
+	err := terraformgenerator.Processor(resourceList)
+
+	// resourceList.Results = &framework.Result{
```
Comment (mengqiy) [inline functions/go/export-terraform/main.go:34]: We should remove these commented lines.

File: functions/go/export-terraform/README.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+## Usage
+
+The function executes as follows:
+
+1. Searches for supported KCC resources in the package
+2. For all KCC resources found, generate the equivalent Terraform
+3. Output the Terraform into a `ConfigMap`.
+
+`export-terraform` function can be executed imperatively as follows:
```
Comment (mengqiy) [inline functions/go/export-terraform/README.md:46]: If the imperative workflow is the only way we want to support, why do we choose to output in a ConfigMap instead of mounting a volume and output the file there?
If desired, we can support both.

File: functions/go/export-terraform/terraformgenerator/terraform_generator_test.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+			outDir := path.Join("..", testDir, tt.Name, "output")
+
+			actualRL, err := testutil.ResourceListFromDirectory(inDir, "")
+			require.NoError(err)
+
+			expectedRL, err := testutil.ResourceListFromDirectory(outDir, "")
+			require.NoError(err)
+
+			// append input items to output
+			expectedRL.Items = append(expectedRL.Items, actualRL.Items...)
```
Comment (mengqiy) [inline functions/go/export-terraform/terraformgenerator/terraform_generator_test.go:59]: Why appending them together?

File: functions/go/export-terraform/terraformgenerator/terraform_generator.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "terraform",
+			Annotations: map[string]string{
+				"config.kubernetes.io/local-config":  "true",
+				"blueprints.cloud.google.com/syntax": "hcl",
+				"blueprints.cloud.google.com/flavor": "terraform",
+				"config.kubernetes.io/index":         "0",
+				"config.kubernetes.io/path":          "terraform.yaml",
+				// TODO: I should't need to insert these, but currently the test framework expects them
```
Comment (mengqiy) [inline functions/go/export-terraform/terraformgenerator/terraform_generator.go:74]: I will look into it

Comment (mengqiy) [reply functions/go/export-terraform/terraformgenerator/terraform_generator.go]: Fixed in #725

**Review by mengqiy — APPROVED** [review]

File: examples/export-terraform-imperative/.expected/diff.patch [hunk 7]
```diff
@@ -1,8 +1,8 @@
-diff --git a/terraform.yaml b/terraform.yaml
+diff --git a/configmap_terraform.yaml b/configmap_terraform.yaml
```
Comment (morgante) [inline examples/export-terraform-imperative/.expected/diff.patch:1]: We should not change this destination path. The contract we've agreed to in go/terraform-krm-export is to place the ConfigMap in `terraform.yaml`.

File: examples/export-terraform-imperative/Kptfile (hunk 2 above)
Comment (morgante) [reply examples/export-terraform-imperative/Kptfile:5]: Yes, I think so. The e2e tests failed when I didn't have a Kptfile.

File: functions/go/export-terraform/README.md (hunk 4 above)
Comment (morgante) [reply functions/go/export-terraform/README.md:46]: A ConfigMap is both more portable (can be served over HTTP APIs easily) and easier to generate (we don't need to give the function filesystem access).

The eventual output which we want for customers is a zip file, but I don't want to entangle zip file generation into this function. See go/terraform-krm-export for more background.

File: examples/export-terraform-imperative/.expected/diff.patch (hunk 7 above)
Comment (mengqiy) [reply examples/export-terraform-imperative/.expected/diff.patch:1]: > // TODO: I should't need to insert these, but currently the test framework expects them

I think I misunderstood your TODO earlier. 
IIUC the function wants to set `internal.config.kubernetes.io/path=terraform.yaml` (`internal.config.kubernetes.io/index` is optional when it's 0), while the legacy annotations (config.kubernetes.io/index and config.kubernetes.io/path) should not be set.
I will update it soon

Comment (mengqiy) [reply examples/export-terraform-imperative/.expected/diff.patch]: Updated.And the CI is passing.

**Comment by mengqiy** [comment]
@morgante PTAL

**Review by morgante — COMMENTED** [review]
LGTM
